dexlib2/src/main/java/org/jf/dexlib2/builder/MethodLocation.java

```diff
@@ -74,167 +73,28 @@ public int getIndex() {
-            }
-            return ImmutableList.of();
-        }
-        return debugItems;
-    }
-
     void mergeInto(@Nonnull MethodLocation other) {
-        if (this.labels != null || other.labels != null) {
-            List<Label> otherLabels = other.getMutableLabels();
-            for (Label label: this.getImmutableLabels()) {
-                label.location = other;
-                otherLabels.add(label);
-            }
-            this.labels = null;
-        }
+        labels.mergeItemsInto(other, other.labels);
 
-        if (this.debugItems != null || other.labels != null) {
-            // We need to keep the debug items in the same order. We add the other debug items to this list, then reassign
-            // the list.
-            List<BuilderDebugItem> debugItems = getMutableDebugItems();
-            for (BuilderDebugItem debugItem: debugItems) {
-                debugItem.location = other;
-            }
-            debugItems.addAll(other.getImmutableDebugItems());
-            other.debugItems = debugItems;
-            this.debugItems = null;
-        }
+        debugItems.mergeItemsInto(other, other.debugItems);
     }
 
     @Nonnull
     public Set<Label> getLabels() {
-        return new AbstractSet<Label>() {
-            @Nonnull
-            @Override public Iterator<Label> iterator() {
-                final Iterator<Label> it = getImmutableLabels().iterator();
-
-                return new Iterator<Label>() {
-                    private @Nullable Label currentLabel = null;
-
-                    @Override public boolean hasNext() {
-                        return it.hasNext();
-                    }
-
-                    @Override public Label next() {
-                        currentLabel = it.next();
-                        return currentLabel;
-                    }
-
-                    @Override public void remove() {
-                        if (currentLabel != null) {
-                            currentLabel.location = null;
-                        }
-                        it.remove();
-                    }
-                };
-            }
-
-            @Override public int size() {
-                return getImmutableLabels().size();
-            }
-
-            @Override public boolean add(@Nonnull Label label) {
-                if (label.isPlaced()) {
-                    throw new IllegalArgumentException("Cannot add a label that is already placed. You must remove " +
-                            "it from its current location first.");
-                }
-                label.location = MethodLocation.this;
-                getMutableLabels().add(label);
-                return true;
-            }
-        };
+        return labels.getModifiableItems(MethodLocation.this);
     }
 
     @Nonnull
     public Label addNewLabel() {
-        Label label = new Label(this);
-        getMutableLabels().add(label);
-        return label;
+        Label newLabel = new Label();
```

Comment:
Because we set the location implicitly in the following line.
Whenever we add a label, we set its location to `this`.

Originally, we had two different locations in the code that set the location of a label to `this`.

We had 
https://github.com/JesusFreke/smali/blob/7079014a29869e1bac22226681de3471b6dc11b5/dexlib2/src/main/java/org/jf/dexlib2/builder/MethodLocation.java#L170
And 
https://github.com/JesusFreke/smali/blob/7079014a29869e1bac22226681de3471b6dc11b5/dexlib2/src/main/java/org/jf/dexlib2/builder/MethodLocation.java#L212